sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java

```diff
@@ -96,21 +137,53 @@ public void setNextRow(T row) {
     public void validateValue(Reference target, Object providedValue) {
         Input<?> input = toValidate.get(target);
         if (input != null) {
-            Object generatedValue = input.value();
-            //noinspection unchecked
-            if (target.valueType().compareValueTo(generatedValue, providedValue) != 0) {
-                throw new IllegalArgumentException(
-                    "Given value " + providedValue +
-                    " for generated column " + target.column() +
-                    " does not match calculation " + ((GeneratedReference) target).formattedGeneratedExpression() + " = " +
-                    generatedValue
-                );
+            doValidateValue(target, input.value(), providedValue);
+        }
+    }
+
+    private void doValidateValue(Reference target, Object generatedValue, Object providedValue) {
+        //noinspection unchecked
+        if (target.valueType().compareValueTo(generatedValue, providedValue) != 0) {
+            throw new IllegalArgumentException(
+                "Given value " + providedValue +
+                " for generated column " + target.column() +
+                " does not match calculation " + ((GeneratedReference) target).formattedGeneratedExpression() + " = " +
+                generatedValue
+            );
+        }
+    }
+
+    void checkNestedValues(Reference target, Object providedValue) {
```

Comment:
I think this lacks some high level explanation *why* nested references have to be treated in a special why.